.eslintrc.json

```diff
@@ -7,6 +7,15 @@
         "sourceType": "module",
         "ecmaVersion": 2020
     },
+    "settings": {
+        "jsdoc": {
+            "mode": "typescript",
+            "preferredTypes": {
```

Comment:
I think we should stick with uppercase Object, as we are doing that in most places already.

Reply:
Yeah, it's currently set as such in `eslint-config-eslint`.

But it gets a little complicated on what to do in this PR, however, and I need to cover some background, especially since the TypeScript doc details are kind of sparse and scattered and can't just refer you there.

Here's a summary of the situation between `object` vs. `Object`:

1. In non-JSDoc TypeScript, the two are different, with `Object` being [discouraged](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#number-string-boolean-symbol-and-object) and being  [treated like `any`](https://github.com/microsoft/TypeScript/issues/15105#issuecomment-350105361) (though testing in the playground, it appears to diverge from `any` in not allowing `null` or `undefined`), whereas `object` [allows](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type) matching anything that is basically `typeof` "object" except for `null`. (Apparently `Object` works with whatever has `Object` on the prototype, thus allowing for strings, numbers, booleans, and symbols too, whereas `object` was [meant](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type) specifically to allow for `Object.create(null)` type objects which do not have Object.)
1. In TypeScript-flavored JSDoc, there is meant to be no difference between the two per https://github.com/microsoft/TypeScript/issues/15105#issuecomment-350114644 (note also the response from the TypeScript team concurring), apparently because they were striving more for JSDoc compatibility than for consistency between TypeScript and TypeScript JSDoc
1. There is an issue ([#20555](https://github.com/microsoft/TypeScript/issues/20555)), reporting that with types like `object<string, number>`, `object` will get treated here as "any" and that issue aims to fix the unintentional differences within JSDoc between `object` and `Object` (even though it would mean less consistency between TypeScript and TypeScript JSDoc).

Based on this, it indeed shouldn't matter which we use in JSDoc in the future, and even now, `Object` may indeed seem preferable, as you suggest, until the issue is fixed, since it can be used in JSDoc.

However, when we automate the building of a declaration file, we get:

1. `Object` (as expected) occurs in the JSDoc (where it is at least not too strict to conflict with the TypeScript below if TypeScript gets preference (the TS playground at least seems to ignore the JSDoc), though it is too liberal in accepting some primitives) (in theory, `Object.create(null)` objects would not work, but the playground does not seem to enforce this)

1. A listing of `node: Object` in TypeScript proper, meaning in this context, that any item will be allowed including numbers and strings--something we don't want to allow here (and is worse than the existing `@types/eslint-visitor-keys` file in this regard).

```js

/**
 * Get visitor keys of a given node.
 * @param {Object} node The AST node to get keys.
 * @returns {readonly string[]} Visitor keys of the node.
 */
export function getKeys(node: Object): readonly string[];

```

By using "object", we instead get `object` in both places, thus avoiding the `node` argument being allowed to be too liberal ("object" ends up in the JSDoc too, meaning it currently would act like "any" there, but again, it appears the TypeScript overrides the JSDoc in TypeScript proper where it can occur.

(For what it's worth, in `eslint-plugin-jsdoc`, we are planning to enforce the behavior in `typescript` mode that by default `object` should be used in JSDoc for greater consistency with TypeScript though for `Object<string, number>` types, `Object` will be required due to the current TypeScript issue. So if you are persuaded, `eslint-config-eslint` could just rely on the default at that time, obviating the need for this extra config I added.)

Reply:
Wow thanks for the detailed explanation. With all that background, I agree it seems like “object” is the better choice. 👍